mapo/agents/mapo/mapo_policy.py

```diff
@@ -0,0 +1,143 @@
+    with tf.compat.v1.variable_scope(Q_SCOPE):
+        policy.q_values, policy.q_preprocessor = build_continuous_q_function(
+            policy.get_obs_input_dict(),
+            policy.observation_space,
+            policy.action_space,
+            policy.config,
+        )
+
+    # Fitted Q loss (using trajectory returns)
+    critic_loss = tf.reduce_mean(
+        tf.square(
+            policy.q_values(batch_tensors[SampleBatch.ACTIONS])
+            - batch_tensors[Postprocessing.ADVANTAGES]
+        )
+    )
+
+    # DPG loss
+    actor_loss = -tf.reduce_mean(policy.q_values(policy.actions))
+
+    return actor_loss + critic_loss
+
+
+def check_action_space(policy, obs_space, action_space, config):
+    # pylint: disable=unused-argument
+    """Check if the action space is suited to DPG."""
+    if not isinstance(action_space, Box):
+        raise UnsupportedSpaceException(
+            "Action space {} is not supported for MAPO.".format(action_space)
+        )
+    if len(action_space.shape) > 1:
+        raise UnsupportedSpaceException(
+            "Action space has multiple dimensions {}.".format(action_space.shape)
+            + "Consider reshaping this into a single dimension, using a Tuple action"
+            "space, or the multi-agent API."
+        )
+
+
+def build_continuous_q_function(input_dict, obs_space, action_space, config):
+    """Construct the state preprocessor and critic layers."""
+    if config["use_state_preprocessor"]:
+        state_preprocessor = ModelCatalog.get_model(
+            input_dict, obs_space, action_space, 1, config["model"]
+        )
+        processed_obs = state_preprocessor.last_layer
+    else:
+        state_preprocessor = None
+        processed_obs = input_dict[SampleBatch.CUR_OBS]
+
+    activation = config["critic_hidden_activation"]
+    hiddens = [
+        keras.layers.Dense(units=hidden, activation=activation)
+        for hidden in config["critic_hiddens"]
+    ]
+    layers = hiddens + [keras.layers.Dense(units=1, activation=None)]
+    model = keras.Sequential(layers=layers)
+
+    def evaluate_actions(actions):
+        return model(tf.concat([processed_obs, actions], axis=1))
+
+    return evaluate_actions, state_preprocessor
+
+
+def build_deterministic_policy(input_dict, obs_space, action_space, config):
+    """Contruct deterministic policy tf graph."""
+    if config["use_state_preprocessor"]:
+        state_preprocessor = ModelCatalog.get_model(
+            input_dict, obs_space, action_space, 1, config["model"]
+        )
+        action_out = state_preprocessor.last_layer
+    else:
+        state_preprocessor = None
+        action_out = input_dict[SampleBatch.CUR_OBS]
+
+    activation = config["actor_hidden_activation"]
+    for hidden in config["actor_hiddens"]:
+        action_out = keras.layers.Dense(units=hidden, activation=activation)(action_out)
+    action_out = keras.layers.Dense(units=action_space.shape[0], activation=None)(
+        action_out
+    )
+
```

Comment:
When we refactor this code, I think it would be interesting to extend `tf.keras.layers.Layer` to define our own `ActionBoundedLayer` so to encapsulate this squashing function (or even to implement other approaches, e.g., clipping, ...). 

That way we could have a proper `tf.keras.models.Sequential` model here the same way function `build_continuous_q_function` builds the critic network...

For this initial version this is fine...